### ISSUE TITLE
Enable OpaqueTensor to possess Storage to allow it to view from CPUTensor

### ISSUE THREAD

**Issue by CaoZhongZ** [pytorch/pytorch#34939]
## 🚀 Feature
Allow OpaqueTensor to have storage. Implement OpaqueTensor<<ideep::tensor>> in a way that use Storage as its buffer manager.

## Motivation

  MKLDNN Layout is a physical extension to stride layout (always stride logically), so it is compatible to CPUTensor when transform from it. The copy nature of to_mkldnn will hurt performance when MKLDNN ops scattered in models, while it could be a view style.

## Pitch

Improve performance when using MKLDNN

## Additional context

  The parameter T of the OpaqueTensor template is the class to describe underlining memory buffer, while it does not have to manage the buffer of its own. If we limit the responsibility of T to meta description and let Storage for resource management, then an Opaque Tensor could naturally view from CPUTensor when they are compatible. Even when Opaque T described a layout that beyond the semantics of a stride Tensor we could still choose to use Storage for buffer management as long as we don’t pass it out to incompatible tensor types.

  A concrete case of this is, as we could see in DNNL Tensor if we use ideep::tensor for meta management and let Storage manage the buffer, then we could make to_mkldnn a view style op instead of a memory copy. In certain cases when DNNL Tensor format is compatible with CPUTensor (nearly all the cases other than some of 4D block format) we could optimize out unnecessary copy in transforming DNNL Tensor from or to CPUTensor.

**Comment by jgong5** [comment]
@CaoZhongZ Are you proposing to extend the semantics of `Storage` such that it could be no longer (offset, strided layout)? Would this BC breaking? Note that the code can get the `Storage` from a `Tensor` and pass it and the corresponding offset along to initialize another `Tensor` assuming strided layout. I guess that is why opaque tensor does not hold `Storage`. Can we make the `Storage` internal only to the opaque tensor just for holding its ref?

Alternatively, we may hold the ref inside ideep::tensor with void* to Storage and a custom deleter (similar to c10::DataPtr).

**Comment by CaoZhongZ** [comment]
I might be naive and optimistic in this:
1. Both 'has_storage' and 'storage' are virtual, which means I could block storage request when format are not compatible.
2. or if has_storage, we use storage, if not, ideep::tensor behavior as descriptor and manager.
Clean and clear.

Hold the ref and deleter in a void pointer? That's too vague and not straight forward.

**Comment by jgong5** [comment]
> or if has_storage, we use storage, if not, ideep::tensor behavior as descriptor and manager.

@CaoZhongZ Yes, this should work. Basically we never use `Storage` to store data with opaque layout, which is safe to fit into current design.

**Comment by ngimel** [comment]
cc @alband for view implications for autograd.

**Comment by albanD** [comment]
I am not super familiar about the OpaqueTensor contruct but from the autograd point of view, when a function is defined in `tools/autograd/derivatives.yaml` like `to_mkldnn` is [here](https://github.com/pytorch/pytorch/blob/1f4a4aaf643b70ebcb40f388ae5226a41ca57d9b/tools/autograd/derivatives.yaml#L923-L924), the signature of the function must reflect properly if the result is a view of the input (see for example how it is done for unsqueeze [here](https://github.com/pytorch/pytorch/blob/1f4a4aaf643b70ebcb40f388ae5226a41ca57d9b/tools/autograd/derivatives.yaml#L938)).
Moreover, it must be added to [this list](https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_autograd.py#L39) as well.

Note that if the function does not always return views, it is trickier.
In that case, you can take an approach similar to `.contiguous()` where the user-facing function is not defined directly in `derivatives.yaml` but internally will call some functions that are defined there (one that creates a view and one that does not).

Let me know if you want more details !

**Comment by CaoZhongZ** [comment]
Thank you for the info. It would indeed always return view if we were done the modification, should backward of the op be view like?

**Comment by albanD** [comment]
There is no restriction on the backward, it can do whichever is most efficient.

Question though: What shouls happen if I do:
```python
a = torch.rand(10, requires_grad=True)
a_mkl = a.to_mkldnn()
a_mkl += 2
# Should a be changed?
# Should the gradients be properly computed?
```

**Comment by CaoZhongZ** [comment]
Current status, 'a' will keep the original content and 'a_mkl' will be a's contents plus 2. After implement the proposal, 'a_mkl' and 'a' will both plus 2.

In this situation, my intension is to_mkldnn must be conforming to 'view' function with same shape of 'a', except the type of 'a_mkl', it is a OpaqueTensor<MKLDNN> now.

**Comment by jgong5** [comment]
> There is no restriction on the backward, it can do whichever is most efficient.

@albanD @CaoZhongZ Any option to get the backward of `to_mkldnn` more efficient? Currently we always use `to_dense` with explicit copy while the input to `to_dense` would be either opaque or strided. In the latter case, `to_dense` would be a view as well.

**Comment by CaoZhongZ** [comment]
@albanD I have a further question about optional view case. Like aten::reshape(Tensor self, int[] shape) -> Tensor, there is no alias mark (a) on self and result tensor. In this case, JIT algorithm won't presume possible alias between self and result, which is an error when reshape actually returns view in some cases.

So I want to know exactly the semantics of the alias mark (a), is it a necessary or sufficient or whatever the definition already have?

**Comment by albanD** [comment]
> After implement the proposal, 'a_mkl' and 'a' will both plus 2.

The problem here is that all the logic that handles view/inplace in the autograd engine assumes that both are strided Tensors. In particular, we need to do fancy stuff with the graph to support this and we will sometimes replace your custom backward function.
In this case, this is problematic because you want a special backward function.

One solution could be to forbid inplace on the result of that view. `a_mkl += 2` would thus raise an error stating that you cannot modify a Tensor send to mkldnn directly inplace when it requires gradients. You must clone it first. Would that work for you?

> there is no alias mark (a) on self and result tensor

From the autograd point of view, you can find more details [here](https://github.com/pytorch/pytorch/blob/781f590f338697705e4690801770e65f1fb69d44/tools/autograd/derivatives.yaml#L748-L753). But the idea is that it is *not* an elementary function of the autograd (there is no ReshapeBackward). And the reshape will either create a graph that contains just a ViewBackward when it returns a view or a CloneBackward + ViewBackward (not sure if these exact Nodes are used but in spirit) when it does a copy.

For the JIT, I don't know how they handle it.

**Comment by CaoZhongZ** [comment]
> In particular, we need to do fancy stuff with the graph to support this and we will sometimes replace your custom backward function.

I poked autograd with some code and learned a lot, you are absolutely right, that's really a lot of fancy stuff there.

> One solution could be to forbid inplace on the result of that view. a_mkl += 2 would thus raise an error stating that you cannot modify a Tensor send to mkldnn directly inplace when it requires gradients. You must clone it first. Would that work for you?

Yes, definitely work for almost all the cases. However I don't want to abandon a small case:
```
a = torch.rand(10, requires_grad=True)
a_mkl = a.to_mkldnn()
a_mkl.relu_()
```
However, if there is no easy solution, we could make it simple, just block in-place case.

**Comment by CaoZhongZ** [comment]
@albanD Hi, I have a hard-time to understand this example. What does it mean?
```
# Have:
#   base.requires_grad = False
#   var.requires_grad = True
base[1] = var  # i.e., base[1].copy_(var)
torch.autograd.grad(base.sum(), var)  <- should return an all ones tensor
```

My experiment (no idea what i'm doing)
```
In [15]: base = torch.zeros(10, requires_grad=False)
In [16]: var = torch.ones(1, requires_grad=True)
In [18]: base[1]=var
In [19]: grad = torch.autograd.grad(base.sum(), var) 
In [20]: grad
Out[20]: (tensor([1.]),)
```

and Next
```
# Have:
#   base.requires_grad = False
#   var.requires_grad = True
view = base[1]
base.copy_(var)
torch.autograd.grad(view.sum(), var)  <- should return a tensor with
                                         var[1] filled with all ones and
                                         zeros everywhere else
```

I got (still no clue)
```
In [2]: base = torch.zeros(10, requires_grad=False)
In [3]: view = base[1]
In [4]: var = torch.ones(1, requires_grad=True)
In [5]: base.copy_(var)
Out[5]: tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.], grad_fn=<CopyBackwards>)
In [6]: torch.autograd.grad(view.sum(), var)
Out[6]: (tensor([1.]),)
```

**Comment by albanD** [comment]
> However I don't want to abandon a small case

The problem with this case is that if the user also modifies `a` inplace, we have to rewrite the graph to be able to compute the right gradients and we won't be able to handle that.
To make this work, we would need to insert in the backward pass multiple calls to `to_mkldnn` and back to regular CPU. This would be very complex.

> I have a hard-time to understand this example. 

The right question to ask yourself to understand what `autograd.grad(out, inp)` does is: By how much `out` will change if I change `inp` a little and rerun my program?
In the first example, you can see that if you move var by `eps`, then the out will move by `eps` as well. So the gradient wrt `inp` is `1`.
For the second one, if you change `var` a by `eps` then `out` will change by `eps` as well, so you get `1` again.

**Comment by ezyang** [comment]
You've identified a problem ("cannot run MKLDNN operations on standard layout cpu tensors without doing a copy first") which I agree with, and a solution ("enable opaquetensor to possess storage") which I disagree with. A much simpler way to solve the problem is to just not use the MKLDNN Opaque Tensor for standard layout tensors, and just use an old fashioned CPU tensor in those cases. You can use the opaque tensor for your block layout situations only.

**Comment by CaoZhongZ** [comment]
> The problem with this case is that if the user also modifies `a` inplace, we have to rewrite the graph to be able to compute the right gradients and we won't be able to handle that.
> To make this work, we would need to insert in the backward pass multiple calls to `to_mkldnn` and back to regular CPU. This would be very complex.

I know, we could block the case. Since we are concerning computation bound ops, they can't be in-place.

**Comment by CaoZhongZ** [comment]
> You've identified a problem ("cannot run MKLDNN operations on standard layout cpu tensors without doing a copy first") which I agree with, and a solution ("enable opaquetensor to possess storage") which I disagree with. A much simpler way to solve the problem is to just not use the MKLDNN Opaque Tensor for standard layout tensors, and just use an old fashioned CPU tensor in those cases. You can use the opaque tensor for your block layout situations only.

Then what is the output when we applied MKLDNN operations on standard layout tensors. Currently we kind have two mode of it: call a MKLDNN operation on standard layout tensors we get a standard layout tensor, call a MKLDNN operation on opaque tensors we could get a opaque tensor which requires a plain copy. You suggest that we call MKLDNN operation on standard layout tensors and output opaque tensor?

**Comment by ezyang** [comment]
The invariant is: standard layout == CPU Tensor; blocked layout == opaque tensor. If you have operations that take standard layout tensors and produce block layout, yes, you will take in a CPU tensor and produce an opaque tensor; and vice versa. Just like we have operations that take in CPU tensors and output CUDA tensors.

**Comment by CaoZhongZ** [comment]
That's exactly the programming model of MKLDNN, I'm up to that. Then my interest shift to, can I get a fallback when an operation does not support Opaque Tensor? Exploiting the new dispatching mechanism?

**Comment by ezyang** [comment]
You mean for pointwise operations? In general none of our kernels are going to know how to deal with non-standard memory layout, so I don't know how you're going to fall back to strided. Do you mean convert to strided, and then run the strided version? There's no easy way to do it, since we tend not to like doing these implicit conversions (it's a hidden performance cliff); for each implicit conversion you want to add, you'll have to program it manually.

**Comment by CaoZhongZ** [comment]
I'm stuck in the status quo. Two path in convolution use DNNL in two different way as I mentioned early, if I change DNNL ops to accept CPU tensor and output opaque tensor then I break one. The only way to contain both usage simultaneously is to do a automatically fall back, which is not good practice.

The status quo can't be accepted when dnnl ops are archipelago in models then there would be many manually added to_mkldnn (copy) to actually hurt performance. That's why I choose to maintain and improve the status quo and do the view style to_mkldnn. At least ops could decide which path to run according to tensor type. I also want to make no change (ideally) in PyTorch autograd. The ideal situation I'm pondering is that we don't do implicit conversion in PyTorch but we get the trigger ready.

Now all these are just in talk, ideas are vague, we like comments and suggestions at least set for a promising path.

**Comment by ezyang** [comment]
I am having a little trouble understanding your comment, so let me see if I have got it correctly. Previously, you said:

>  Currently we kind have two mode of it: call a MKLDNN operation on standard layout tensors we get a standard layout tensor, call a MKLDNN operation on opaque tensors we could get a opaque tensor which cost a copy to do the transfer. 

If you have an MKLDNN operation that takes standard tensor and produces standard tensor, what I am suggesting is that you *never* output an opaque tensor. In comes a CPU tensor, out goes a CPU tensor; MKLDNN kernel is just something we transparently call to do a faster implementation.

> That's why I choose to maintain and improve the status quo and do the view style to_mkldnn.

There are a few reasons why I don't want to do this.

First, it introduces another way of representing a tensor which is equivalent to an existing representation we already have: CPU tensor versus standard layout MKLDNN tensor. The Python philosophy is that there should only be one way to do it. We already have a way to do it, let's use it.

Second, your intended implementation mechanism is to make use of virtual methods on TensorImpl. We have a longstanding issue to devirtualize these methods on TensorImpl, as they are only used by XLA, for performance reasons. Adding more virtual users makes it harder to deal with this case. It wouldn't be so bad if there was a good reason for doing it, but as mentioned above, it doesn't seem like this use case is actually justified.

**Comment by CaoZhongZ** [comment]
> The Python philosophy is that there should only be one way to do it. We already have a way to do it, let's use it.

Got it. Then I think a solution do not involve copy could be like
```
a.to_mkldnn() #<— dispatch id add MkldnnCPUTensorId, still normal TensorImpl
b = conv2->forward(a) # Allow to return Opaque by checking MkldnnCPUTensorId
```

> as they are only used by XLA, for performance reasons. Adding more virtual users makes it harder to deal with this case. It wouldn't be so bad if there was a good reason for doing it, but as mentioned above, it doesn't seem like this use case is actually justified.

No, I didn't intend to add more I just want to use what I can have, and of course for performance reason, too. I may come back to think about this part when settle the first issue.

**Comment by ezyang** [comment]
> `a.to_mkldnn() #<— dispatch id add MkldnnCPUTensorId, still normal TensorImpl`

I don't think you need to add a dispatch id at all. Just add an if statement inside of the cpu kernels to go to mkldnn :)

**Comment by CaoZhongZ** [comment]
Ah, a little longer story. Since we are at opposite time zone, I should put more information.

There are two cases I have to maintain, for a toy module like:

```
class SomeModule(nn.Module):
    self.conv1 = conv3x3(inplanes, planes, stride)
    self.bn1 = norm_layer(planes)
    self.relu = nn.ReLU(inplace=True)
    self.conv2 = conv3x3(planes, planes)
    self.bn2 = norm_layer(planes)
```
Use case 1:
```
# a is CPU Tensor
b = some_module->forward(a)    #accelerated by DNNL, outputs of convolutions are always transform back to CPU
```

Use case 2:
```
# a is CPU Tensor
a1 = a.to_mkldnn()                         # COPY ALERT !!
b = some_module->forward(a1)   # accelerated by DNNL, outputs of convolutions are (most likely) Opaque tensor
```

I call first case the standard mode, second turbo mode, both path go through DNNL convolution kernel https://github.com/pytorch/pytorch/blob/bd604cb5b7ae4f6388aca461891d620b0d485fbb/aten/src/ATen/native/Convolution.cpp#L689-L696.  If we run standard mode, convolution output will always be stride Tensor, we get DNNL acceleration. if we run turbo mode, we get full acceleration (possible 20%+ even 50%) but sometimes we may face situations that ops do not support Opaque Tensor. Users have to adapt models by adding to_mkldnn/to_dense in this case. 
https://github.com/pytorch/pytorch/blob/bd604cb5b7ae4f6388aca461891d620b0d485fbb/aten/src/ATen/native/mkldnn/Conv.cpp#L125-L130
A very awkward situation exists that due to the copy nature of to_mkldnn, turbo mode can be slower than standard mode when user add too much to_mkldnn.

Now come to our discussion. We need a flag to distinguish turbo mode from standard mode, zero copy to_mkldnn or global switch, anything zero overhead, free of copy. Ideally similar to old one etc. If we input CPUTensor and output Opaque Tensor then we destroyed the first case.

**Comment by ezyang** [comment]
> If we input CPUTensor and output Opaque Tensor then we destroyed the first case.

I still do not understand under what operators you would take a CPU tensor and output an opaque tensor, besides operations which are explicitly trying to get you into the MKLDNN tensor.

Here is an analogous situation: as part of @VitalyFedyunin's work, we are adding support for NHWC layout tensors, orthogonal to NCHW tensors. In analogy to your terminology, let's call NCHW "standard mode" and NHWC "turbo mode".  Similar to your situation, if you are constantly swapping between NCHW layout and NHWC you will pay a performance cost. Our general contract to the users is that if they give us an NCHW tensor, they will get back out an NCHW tensor; similarly, for NHWC, if they give us an NHWC tensor, we will give them NHWC tensor in the end. If they convert, they explicitly ask us to convert. It's important that the user can maintain a mental model of what layout they are in without needing to know the intricate details of the code, as the layout, as you say, has implications on performance.

So here's my question for you: why can't MKLDNN follow these rules? If when given standard layout, if MKLDNN always produces standard layout, then your problem doesn't exist. And if you have some operations where it would be profitable to modify layout while the operation is occurring, we could support this by passing a `layout=` option (in analogy to the `memory_format=` which has already been added to many operations). What is left?!

**Comment by CaoZhongZ** [comment]
That's neat, I take 'layout=' and try fitting some cases. Thanks!
